### ISSUE TITLE
esr128: neqo-crypto and other security code fails to build as we don't meet NSS requirement

### ISSUE THREAD

**Issue by kenz-gelsoft** [kenz-gelsoft/gecko-dev#15]
security/certverifier
```
 0:14.69 In file included from Unified_cpp_certverifier0.cpp:20:
 0:14.69 /boot/home/src/gecko-dev/security/certverifier/NSSCertDBTrustDomain.cpp:1348:23: error: use of undeclared identifier 'PK11_ReadDistrustAfterAttribute'
 0:14.69       SECStatus srv = PK11_ReadDistrustAfterAttribute(
 0:14.69                       ^
 0:15.36 1 error generated.
 0:15.39 /boot/home/src/gecko-dev/config/rules.mk:688: recipe for target 'Unified_cpp_certverifier0.o' failed
 0:15.39 make[4]: *** [Unified_cpp_certverifier0.o] Error 1
 0:15.39 /boot/home/src/gecko-dev/config/recurse.mk:72: recipe for target 'security/certverifier/target-objects' failed
 0:15.39 make[3]: *** [security/certverifier/target-objects] Error 2
 0:21.26 In file included from Unified_cpp_protocol_http1.cpp:83:
 0:21.26 /boot/home/src/gecko-dev/netwerk/protocol/http/Http2Session.cpp:4214:14: error: use of undeclared identifier 'ssl_kea_ecdh_hybrid'; did you mean 'ssl_kea_ecdh_psk'?
 0:21.26   if (kea == ssl_kea_ecdh_hybrid && !StaticPrefs::security_tls_enable_kyber()) {
 0:21.26              ^~~~~~~~~~~~~~~~~~~
 0:21.26              ssl_kea_ecdh_psk
 0:21.26 /packages/nss-3.73.1-1/.self/develop/headers/nss/sslt.h:82:5: note: 'ssl_kea_ecdh_psk' declared here
 0:21.26     ssl_kea_ecdh_psk = 5,
 0:21.26     ^
 0:21.31 In file included from Unified_cpp_protocol_http1.cpp:83:
 0:21.32 /boot/home/src/gecko-dev/netwerk/protocol/http/Http2Session.cpp:4220:58: error: use of undeclared identifier 'ssl_kea_ecdh_hybrid'; did you mean 'ssl_kea_ecdh_psk'?
 0:21.32   if (kea != ssl_kea_dh && kea != ssl_kea_ecdh && kea != ssl_kea_ecdh_hybrid) {
 0:21.32                                                          ^~~~~~~~~~~~~~~~~~~
 0:21.32                                                          ssl_kea_ecdh_psk
 0:21.32 /packages/nss-3.73.1-1/.self/develop/headers/nss/sslt.h:82:5: note: 'ssl_kea_ecdh_psk' declared here
 0:21.32     ssl_kea_ecdh_psk = 5,
 0:21.32     ^
 0:23.88 2 errors generated.
```

neqo-crypto
```
 0:37.23    Compiling neqo-crypto v0.7.9 (https://github.com/mozilla/neqo?tag=v0.7.9#121fe683)
 0:37.55 error[E0425]: cannot find value `ssl_grp_kem_xyber768d00` in module `ssl::SSLNamedGroup`
 0:37.56   --> /boot/home/src/gecko-dev/third_party/rust/neqo-crypto/src/constants.rs:65:35
 0:37.56    |
 0:37.56 65 |         TLS_GRP_KEM_XYBER768D00 = ssl_grp_kem_xyber768d00,
 0:37.57    |                                   ^^^^^^^^^^^^^^^^^^^^^^^ not found in `ssl::SSLNamedGroup`
 0:37.57 error[E0425]: cannot find value `SSL_ENABLE_GREASE` in module `SSLOption`
 0:37.57   --> /boot/home/src/gecko-dev/third_party/rust/neqo-crypto/src/ssl.rs:72:40
 0:37.57    |
 0:37.57 72 |             Self::Grease => SSLOption::SSL_ENABLE_GREASE,
 0:37.57    |                                        ^^^^^^^^^^^^^^^^^ not found in `SSLOption`
 0:37.58 error[E0425]: cannot find value `SEC_OID_XYBER768D00` in module `p11::SECOidTag`
 0:37.58    --> /boot/home/src/gecko-dev/third_party/rust/neqo-crypto/src/lib.rs:127:33
 0:37.58     |
 0:37.58 127 |                 p11::SECOidTag::SEC_OID_XYBER768D00,
 0:37.58     |                                 ^^^^^^^^^^^^^^^^^^^ not found in `p11::SECOidTag`
 0:37.58 error[E0425]: cannot find value `SEC_OID_XYBER768D00` in module `p11::SECOidTag`
 0:37.58    --> /boot/home/src/gecko-dev/third_party/rust/neqo-crypto/src/lib.rs:182:33
 0:37.58     |
 0:37.59 182 |                 p11::SECOidTag::SEC_OID_XYBER768D00,
 0:37.59     |                                 ^^^^^^^^^^^^^^^^^^^ not found in `p11::SECOidTag`
 0:38.03 For more information about this error, try `rustc --explain E0425`.
 0:38.04 error: could not compile `neqo-crypto` (lib) due to 4 previous errors
```

**Comment by kenz-gelsoft** [comment]
neqo is for QUIC, can we disable it for now?

- disable-encryption feature
- like this:
  - https://searchfox.org/mozilla-esr128/source/netwerk/socket/neqo_glue/Cargo.toml#36
- as far as searched code, just commenting out should work too, now

---

- https://github.com/mozilla/neqo/commit/fbc5d62697c968dec968941d4c495f4fa6869c9d
  - requires NSS 3.97+
  - tagged from 0.6.9
- https://github.com/mozilla/neqo/commit/28679859dbcaf91ec1996670a2e7bc95b5c92c5e
  - for this, it seems just revert this may work
  - tagged from 0.6.5
- https://github.com/mozilla/neqo/commit/4142ac6592c6b05b35bba83ecb2fb95d3a9d5737
  - also
  - tagged from 0.7.9

---

Try downgrading to 0.6.4

**Comment by kenz-gelsoft** [comment]
Changing dependency crate version needs a bit work.

https://hg.mozilla.org/releases/mozilla-esr128/rev/30a3144e9358e4abaae7f48a4bfd1c11e65e065d

Try just commenting out for now.

**Comment by kenz-gelsoft** [comment]
> Try downgrading to 0.6.4

This seems not to work (as it became patch.unused)
```diff
~/src/gecko-dev> git diff
diff --git a/Cargo.lock b/Cargo.lock
index d160c1f4949a..d5e10bc0fae1 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -7099,3 +7099,8 @@ dependencies = [
  "crossbeam-utils",
  "flate2",
 ]
+
+[[patch.unused]]
+name = "neqo-crypto"
+version = "0.6.4"
+source = "git+https://github.com/mozilla/neqo?rev=80db3a01f3273c7e742ba560fa99246fc8b30c4f#80db3a01f3273c7e742ba560fa99246fc8b30c4f"
diff --git a/Cargo.toml b/Cargo.toml
index 8591754712fa..1d4c51964026 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -240,3 +240,4 @@ gpu-descriptor = { git = "https://github.com/zakarumych/gpu-descriptor", rev = "
 # Patch for porting to Haiku OS.
 glslopt = { git = "https://github.com/kenz-gelsoft/glslopt-rs", rev = "f9fb33cdab7d55ee0b06fbde64f61be74bd4ef2b" }
 lmdb-rkv-sys = { package = "lmdb-rkv-sys", git = "https://github.com/kenz-gelsoft/lmdb-rs", rev = "c246a7b064a067feaf9d611c20b418af261477d0" }
+neqo-crypto = { package = "neqo-crypto", git = "https://github.com/mozilla/neqo", rev = "80db3a01f3273c7e742ba560fa99246fc8b30c4f" }
```

**Comment by kenz-gelsoft** [comment]
porting newer NSS may be easier

let's look patches in haikupots

---

https://github.com/haikuports/haikuports/blob/master/dev-libs/nss/patches/nss-3.73.1.patchset

 
recent version bump on haikuports required no any additional change. so both importing patchset in-tree and updating port should work easily.

**Comment by kenz-gelsoft** [comment]
BeOS support of NSS removed between esr102 to esr115

https://phabricator.services.mozilla.com/D37215

[D37215.1723043856.patch](https://github.com/user-attachments/files/16512303/D37215.1723043856.patch)

**Comment by kenz-gelsoft** [comment]
About bring back BeOS/Haiku port, I made https://github.com/kenz-gelsoft/gecko-dev/issues/16 .

**Comment by kenz-gelsoft** [comment]
Maybe just because patches for those not applied
```
 0:05.39 /boot/home/src/gecko-dev/gfx/angle/checkout/src/common/system_utils_posix.cpp:204:22: error: use of undeclared identifier 'RTLD_NOLOAD'
 0:05.40         extraFlags = RTLD_NOLOAD;
 0:05.40                      ^
```
```
 0:10.96 In file included from Unified_cpp_ipc_chromium0.cpp:110:
 0:10.96 /boot/home/src/gecko-dev/ipc/chromium/src/base/process_util_posix.cc:145:15: error: use of undeclared identifier 'kSystemDefaultMaxFds'
 0:10.96     max_fds = kSystemDefaultMaxFds;
 0:10.96               ^
 0:10.99 /boot/home/src/gecko-dev/ipc/chromium/src/base/process_util_posix.cc:153:25: error: unknown type name 'kFDDir'
 0:10.99   DirReaderPosix fd_dir(kFDDir);
 0:10.99                         ^
```
```
 0:19.55 In file included from Unified_cpp_js_src_util0.cpp:47:
 0:19.55 /boot/home/src/gecko-dev/js/src/util/NativeStack.cpp:151:3: error: use of undeclared identifier 'pthread_getattr_np'
 0:19.55   pthread_getattr_np(thread, &sattr);
 0:19.55   ^
```
```
 4:05.11 In file included from Unified_cpp_js_src9.cpp:47:
 4:05.11 /boot/home/src/gecko-dev/js/src/threading/posix/PosixThread.cpp:119:8: error: use of undeclared identifier 'pthread_setname_np'; did you mean 'pthread_attr_get_np'?
 4:05.11   rv = pthread_setname_np(pthread_self(), name);
 4:05.12        ^~~~~~~~~~~~~~~~~~
 4:05.12        pthread_attr_get_np
 4:05.12 /boot/system/develop/headers/bsd/pthread.h:21:12: note: 'pthread_attr_get_np' declared here
 4:05.12 extern int pthread_attr_get_np(pthread_t thread, pthread_attr_t* attr);
 4:05.12            ^
 4:05.12 In file included from Unified_cpp_js_src9.cpp:47:
 4:05.13 /boot/home/src/gecko-dev/js/src/threading/posix/PosixThread.cpp:119:43: error: cannot initialize a parameter of type 'pthread_attr_t *' (aka '_pthread_attr **') with an lvalue of type 'const char *'
 4:05.13   rv = pthread_setname_np(pthread_self(), name);
 4:05.13                                           ^~~~
 4:05.13 /boot/system/develop/headers/bsd/pthread.h:21:66: note: passing argument to parameter 'attr' here
 4:05.13 extern int pthread_attr_get_np(pthread_t thread, pthread_attr_t* attr);
 4:05.13                                                                  ^
 4:05.14 In file included from Unified_cpp_js_src9.cpp:47:
 4:05.14 /boot/home/src/gecko-dev/js/src/threading/posix/PosixThread.cpp:129:8: error: use of undeclared identifier 'pthread_getname_np'
 4:05.14   rv = pthread_getname_np(pthread_self(), nameBuffer, len);
 4:05.14        ^
 4:06.26 3 errors generated.
```

**Comment by kenz-gelsoft** [comment]
Fixed with #14

**Comment by kenz-gelsoft** [comment]
Will be fixed by https://github.com/kenz-gelsoft/gecko-dev/pull/17 too.
